### PR TITLE
Auth method

### PR DESCRIPTION
**Pull request by jmauerhan**
- Changed Resource to require RestCient instead of ClientInterface
- Changed RestClient to pass self to Resource instead of child
  ClientInterface
- Added request method in RestClient to enable other client
  implementations
- Added authorization headers in request method in RestClient.
- Updated all tests

Ref #43

### CODE REVIEW

File: src/RestClient.php [hunk 1]
```diff
@@ -144,18 +148,46 @@ public function getResource($resource)
...
     {
         return $this->client;
     }
+
+    /**
+     * @param string $method HTTP method.
+     * @param string|\Psr\Http\Message\UriInterface $uri URI object or string.
+     * @param array $options Request options to apply.
+     *
+     * @return ResponseInterface
```
Comment (pedrochaves) [inline src/RestClient.php]: It's missing the full namespace here.

File: src/RestClient.php [hunk 2]
```diff
@@ -144,18 +148,46 @@ public function getResource($resource)
...
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function request($method, $uri, $options = [])
+    {
+        if ($this->authorization !== null) {
+            $authHeader = sprintf('%s %s', $this->authorization->getAuthType(), $this->authorization->getCredentials());
+            $headers = isset($options['headers']) ? $options['headers'] : [];
+            $headers[] = ['Authorization' => $authHeader];
```
Comment (pedrochaves) [inline src/RestClient.php]: Shouldn't this be `$headers['Authorization'] = $authHeader`?

File: tests/unit/RestClientTest.php [hunk 3]
```diff
@@ -109,4 +101,43 @@ public function testClientAddsCacheMiddlewareToHandlerStackWhenNoCustomClientPro
...
+
+        $customClient = $this->createMock('GuzzleHttp\ClientInterface');
+        $customClient->expects($this->once())
+            ->method('request')
+            ->with($method, $uri, $expectedOptions);
+
+        $restClient = $this->getRestClient(['custom' => $customClient]);
+        $restClient->setAuthorization($auth);
+        $restClient->request($method, $uri, $options);
+    }
```
Comment (pedrochaves) [inline tests/unit/RestClientTest.php]: Very nice! The only thing here would be change from PHPUnit's default mock method to [Prophecy](https://phpunit.de/manual/current/en/test-doubles.html#test-doubles.prophecy), which is already bundled here.

File: src/RestClient.php (hunk 1 above)
Comment (jmauerhan) [inline src/RestClient.php]: It's in a use statement, so it's not required for it to work in most IDE - Is it part of your code style that you require it?

File: src/RestClient.php (hunk 2 above)
Comment (jmauerhan) [inline src/RestClient.php]: Yeah, I think you're right - I'll change it as well.

File: src/RestClient.php (hunk 1 above)
Comment (lucasmezencio) [inline src/RestClient.php]: @pedrochaves as of `ResponseInterface` is imported on top of this file, I think this change has not need to be made.

Comment (pedrochaves) [inline src/RestClient.php]: Ok, then :+1:

**Comment by jmauerhan** [comment]
@pedrochaves the changes requested have been made!

**Comment by lucasmezencio** [comment]
@jmauerhan can you resolve these conflicts before we merge this PR, pls?

Tks. :)

**Comment by lucasmezencio** [comment]
Closed by 09c6fe79bbb182c7b4dc28b36f7c8f7231c37edb. o.O

**Comment by lucasmezencio** [comment]
@jmauerhan your PR was accepted and merged. In someway, the Github does not understood the merge. But it was done.

**Pull request closed without merging**